### PR TITLE
Introduce the Mixer IL, interpreter and compiler.

### PR DESCRIPTION
**Pull request by ozevren**
This CL introduces Mixer IL, interpreter and compiler.

The Mixer IL is an intermediate-language, geared towards efficient processing of incoming attributes to Mixer. It is accompanied by an interpreter that takes the IL based programs and attribute bags to perform evaluations. There is also a compiler that can convert Mixer expressions to IL based programs.

---
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/793)

### CODE REVIEW

**Comment by douglas-reid** [comment]
Should this go on a branch for evaluation before we decide to mainline it ?

**Comment by ozevren** [comment]
I can do that. ( I don't think the instructions mentions branches/evalution
though, or I missed it).

"il" branch?

**Comment by kyessenov** [comment]
is there a design doc for the IL?

**Comment by ozevren** [comment]
There is package level write-up on the IL/instructions and interpreter in
code.

**Comment by douglas-reid** [comment]
@ozevren maybe: `il-experiment` or `mixer-il` for the branch ?

We probably haven't documented this workflow, because we haven't really encountered it yet :). For introducing a component this large and potentially complex, I think we want to keep it separate for a bit. I think there is likely going to be a fair amount of discussion around adding an IL.

**Comment by ozevren** [comment]
Ok. I'll go with mixer-il.

This is my first attempt at contribution in a long while, so do please let
me know if you guys want me go through some steps. I am happy to go through
them, just let me know. :)

**Comment by ZackButcher** [comment]
Yeah, it'd be nice to see a design - the package comments describe "what" but not "why". I'd also really like to see this split up into multiple, incremental PRs (or at least multiple commits so each commit can be looked at independently). It's impossible to review this much code and actually find the problems.

**Comment by ozevren** [comment]
Will do. Should I still send the CLs, or do you want to take a look at a
design doc first?

**Comment by ZackButcher** [comment]
Doesn't matter to me - maybe split things up so we can start the reviews first, since they'll take a while, and produce the design while you're blocked on reviews?

**Comment by mandarjog** [comment]
@ozevren I think the best way to go about it to open a "mixer issue" as
"Proposal: Create Mixer IL, interpreter and compiler for expression language".
You can answer the why / how questions there.
Thanks.

File: pkg/il/interpreter/extern.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+	pTypes []il.Type
+	rType  il.Type
+
+	v reflect.Value
+}
+
+// ExternFromFn creates a new reflection based Extern, based on the given function. It panics if the
+// function signature is incompatible to be an extern.
+//
+// A function can be extern under following conditions:
```
Comment (geeknoid) [inline pkg/il/interpreter/extern.go]: under the following

File: pkg/il/interpreter/extern.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+// - Input parameter types are one of the supported types: string, bool, int64, float64, map[string]string.
```
Comment (geeknoid) [inline pkg/il/interpreter/extern.go]: We don't support time.Duration, time.Time, and []byte?

File: pkg/il/interpreter/extern.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+	outs := e.v.Call(ins)
+	var rv reflect.Value
+	switch len(outs) {
+	case 1:
+		if e.rType != il.Void {
+			rv = outs[0]
+			break
+		}
+		i := outs[0].Interface()
+		if i != nil {
```
Comment (geeknoid) [inline pkg/il/interpreter/extern.go]: Combine with previous line.

File: pkg/il/interpreter/extern.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+			return 0, 0, i.(error)
+
+		}
+	case 2:
+		rv = outs[0]
+		i := outs[1].Interface()
+		if i != nil {
```
Comment (geeknoid) [inline pkg/il/interpreter/extern.go]: Combine with previous line.

File: pkg/il/interpreter/extern.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+	}
+
+	switch e.rType {
+	case il.String:
+		str := rv.String()
+		id := s.GetID(str)
+		return id, 0, nil
+
+	case il.Bool:
+		b := rv.Bool()
```
Comment (geeknoid) [inline pkg/il/interpreter/extern.go]: Simplify:

if rb.Bool() {
  return 1, 0, nil
}
return 0, 0, nil

File: pkg/il/read.go [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+// ReadText parses the given il assembly text and converts it into a Program.
+func ReadText(text string) (Program, error) {
+	p := NewProgram()
+	err := MergeText(text, p)
+	if err != nil {
```
Comment (geeknoid) [inline pkg/il/read.go]: Combine with previous line

File: pkg/il/builder.go [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+	f.op0(Lookup)
+}
+
+// ILookup appends the "ilookup" instruction to the byte code.
+func (f *Builder) ILookup(v string) {
+	f.op1(ILookup, f.id(v))
+}
+
+// AllocateLabel allocates a new label value for use within the code.
+func (f *Builder) AllocateLabel() string {
```
Comment (ozevren) [inline pkg/il/builder.go]: You may need to reference a label before you need to declare it. i.e.:
jmp Foo
...
Foo:

In that case, you need to allocate the label first, use it in jmp and then set its position.

File: pkg/il/function.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package il
+
+// Function contains metadata about an IL-based function that is implemented in a program.
+type Function struct {
+	// Name is the name of the function. It needs to be globally unique.
+	Name string
```
Comment (ozevren) [inline pkg/il/function.go]: Interesting. Yes that works. Thanks for the suggestion.

File: pkg/il/generate.sh [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+// =================================================================================================
+// WARNING! This file is auto-generated. Do not hand-edit. Instead, edit interpreterRun.got
```
Comment (ozevren) [inline pkg/il/generate.sh]: It is not a valid .go file and makes editing really hard when intellisense barfs due to preprocessor directives. 
I can still rename it if this is big concern though.

File: pkg/il/interpreter/extern.go (hunk 2 above)
Comment (ozevren) [inline pkg/il/interpreter/extern.go]: Not yet? We can add support for these. time.Duration is int64, so effectively we can already support it. time.Time seems a bit heavy weight (64bits + 32bits + pointer), which needs to be kept in "heap". Similarly for bytes.

**Pull request closed without merging**